### PR TITLE
added min_length to LearnRequest

### PR DESCRIPTION
**Pull request by carolina-museum**
#95 
I added constraints on input to LearnRequest

It gives me 
`fastapi     | INFO:     192.168.65.1:45565 - "POST /learn HTTP/1.1" 422 Unprocessable Entity`
when I feed short sequences such as "dog" (less than 4 letters)

### CODE REVIEW

**Comment by huyfififi** [comment]
can you try to add tests to confirm the validation works? 😃  you can refer to the test I have added in this repository.

**Comment by carolina-museum** [comment]
ok! I will try that too!

**Comment by huyfififi** [comment]
FYI we should test boundary conditions (I'm not sure this terminology is correct).
For this pull request, we should at least check the following scenarios.
- `len(text) = 3` => validation error
- `len(text) = 4` => success

**Comment by huyfififi** [comment]
and by having the validation, the comment we have added in the previous pull request is no longer needed.
It is recommended to keep information in only one place to prevent inconsistencies.

**Comment by huyfififi** [comment]
> and by having the validation, the comment we added to the previous pull request is no longer needed. It is recommended to keep information in only one place to prevent inconsistencies, i.e., by having the comment and the validation at the same time, the comment may be updated, but the min value stays the same, confusing collaborators.

on second thought, it is a bit more complicated than I expected. Let's skip this for now.

**Comment by carolina-museum** [comment]
> # pytest test_LearnRequest.py
> ============================================================== test session starts ==============================================================
> platform linux -- Python 3.12.7, pytest-8.3.3, pluggy-1.5.0
> rootdir: /app
> configfile: pyproject.toml
> plugins: anyio-4.6.0
> collected 4 items                                                                                                                               
> 
> test_LearnRequest.py F...                                                                                                                 [100%]
> 
> =================================================================== FAILURES ====================================================================
> ______________________________________________________________ test_LearnRequest1 _______________________________________________________________
> 
> client = <starlette.testclient.TestClient object at 0xffff8a6a76e0>
> 
>     def test_LearnRequest1(client: TestClient):
>         response = client.post(
>             "/learn",
>             json={
>                 "text": "dog",
>             },
>         )
> >       assert response.status_code == 204
> E       assert 422 == 204
> E        +  where 422 = <Response [422 Unprocessable Entity]>.status_code
> 
> test_LearnRequest.py:11: AssertionError
> ============================================================ short test summary info ============================================================
> FAILED test_LearnRequest.py::test_LearnRequest1 - assert 422 == 204
> ========================================================== 1 failed, 3 passed in 1.82s ==========================================================

File: app/tests/api/routes/test_LearnRequest.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    response = client.post(
+        "/learn",
+        json={
+            "text": "The family on my father's side is descended from Caspar Keller, a native of Switzerland, who settled in Maryland.",
+        },
+    )
+    assert response.status_code == 204
+
+
+def test_LearnRequest4(client: TestClient):
```
Comment (huyfififi) [inline app/tests/api/routes/test_LearnRequest.py:34]: although we must try to keep the main code base as minimal as possible, it is totally fine to have redundant test cases.
but just for your information, the third test case and the fourth test case covers the same scenario (len(text) > 4) and some might argue that it is redundant

File: app/tests/api/routes/test_LearnRequest.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+from fastapi.testclient import TestClient
+
+
+def test_LearnRequest1(client: TestClient):
```
Comment (huyfififi) [inline app/tests/api/routes/test_LearnRequest.py:4]: according to PEP8 https://peps.python.org/pep-0008/#function-and-variable-names function names should be snake_case

File: app/tests/api/routes/test_LearnRequest.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    response = client.post(
+        "/learn",
+        json={
+            "text": "dog",
+        },
+    )
+    assert response.status_code == 204
+
+
+def test_LearnRequest2(client: TestClient):
```
Comment (huyfififi) [inline app/tests/api/routes/test_LearnRequest.py:14]: Please rename functions with more descriptive ones. Collaborators should be able to understand what scenario the tests cover by looking at their names

File: app/tests/api/routes/test_LearnRequest.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def test_LearnRequest1(client: TestClient):
+    response = client.post(
+        "/learn",
+        json={
+            "text": "dog",
+        },
+    )
+    assert response.status_code == 422
```
Comment (huyfififi) [inline app/tests/api/routes/test_LearnRequest.py:11]: please assert the response body, too

File: app/tests/api/routes/test_LearnRequest.py (hunk 1 above)
Comment (carolina-museum) [reply app/tests/api/routes/test_LearnRequest.py:34]: I got it! Thank you!
For this, I was thinking that the third one is one paragraph and the fourth one is several paragraphs

Comment (huyfififi) [reply app/tests/api/routes/test_LearnRequest.py:34]: plz contain the information in the test name 😄  it would be super helpful

File: app/tests/api/routes/test_LearnRequest.py (hunk 3 above)
Comment (carolina-museum) [reply app/tests/api/routes/test_LearnRequest.py:14]: Ah I see, that's how we describe it better! I will go and make changes
Thank you!

File: app/tests/api/routes/test_LearnRequest.py (hunk 1 above)
Comment (carolina-museum) [reply app/tests/api/routes/test_LearnRequest.py:34]: Sounds good!

File: app/tests/api/routes/test_LearnRequest.py (hunk 4 above)
Comment (carolina-museum) [reply app/tests/api/routes/test_LearnRequest.py:11]: What is it to assert the response body?
Would it be like if the json file exists? like: `assert response.json()`

Comment (huyfififi) [reply app/tests/api/routes/test_LearnRequest.py:11]: When using the HTTP protocol to communicate between web servers and user clients, JSON is a common choice for transferring or representing data. Yes, `response.json()` should return error messages so that user clients (such as web browsers) can see what happened. You might want to

```
assert response.json() == {...content...}
```

**Comment by huyfififi** [comment]
Just in case you are unfamiliar with HTTP status codes https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

File: app/tests/api/routes/test_LearnRequest.py (hunk 4 above)
Comment (carolina-museum) [reply app/tests/api/routes/test_LearnRequest.py:12]: I see🤔 I tried this, but is this what you are trying to say...?
I don't think I understood what I need to do🤔

Comment (huyfififi) [reply app/tests/api/routes/test_LearnRequest.py:12]: Yes, the newly added assertion is correct! This is the error response when we send the invalid requests from Postman or other tools.

File: app/tests/api/routes/test_LearnRequest.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+def test_learn_request_right_on_inclusive_limit(client: TestClient):
+    response = client.post(
+        "/learn",
+        json={
+            "text": "cats",
+        },
+    )
+    assert response.status_code == 204
+
+    # assert response.json() == {"detail": [
```
Comment (huyfififi) [inline app/tests/api/routes/test_LearnRequest.py:36]: Let's remove debugging code such as this comment section and the print() at line 11 👍

**Comment by carolina-museum** [comment]
> Just in case you are unfamiliar with HTTP status codes https://developer.mozilla.org/en-US/docs/Web/HTTP/Status When creating embeddings in database, I think returning 201 Created along with some insertion results is more appropriate, but we can fix it later

Thank you for the information! I will look into them

File: app/tests/api/routes/test_LearnRequest.py (hunk 5 above)
Comment (carolina-museum) [reply app/tests/api/routes/test_LearnRequest.py]: removed👍

**Review by huyfififi — APPROVED** [review]
✅ 

do you think you can tackle these? #50 #51 🥹

**Comment by carolina-museum** [comment]
Thank you so much!!!
I'll try #50 and #51 :)
